axum-extra/src/extract/cookie/mod.rs

```diff
@@ -95,26 +95,41 @@ where
     type Rejection = Infallible;
 
     async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
-        let mut jar = cookie::CookieJar::new();
-        for cookie in cookies_from_request(req) {
-            jar.add_original(cookie);
-        }
-        Ok(Self { jar })
+        Ok(Self::from(req.headers()))
     }
 }
 
-fn cookies_from_request<B>(
-    req: &mut RequestParts<B>,
-) -> impl Iterator<Item = Cookie<'static>> + '_ {
-    req.headers()
+fn cookies_from_request(headers: &HeaderMap) -> impl Iterator<Item = Cookie<'static>> + '_ {
+    headers
         .get_all(COOKIE)
         .into_iter()
         .filter_map(|value| value.to_str().ok())
         .flat_map(|value| value.split(';'))
         .filter_map(|cookie| Cookie::parse_encoded(cookie.to_owned()).ok())
 }
 
+impl<'a> From<&'a HeaderMap> for CookieJar {
+    fn from(headers: &'a HeaderMap) -> Self {
+        let mut jar = cookie::CookieJar::new();
+        for cookie in cookies_from_request(headers) {
+            jar.add_original(cookie);
+        }
+        Self { jar }
+    }
+}
+
 impl CookieJar {
+    /// Create a new empty `CookieJar` from a map of request headers.
```

Comment:
I prefer the new function with a parameter (as documented here, but not as implemented) over the `From` impl because that is harder to discover. The `Default` implementation probably also makes sense for some cases, maybe its existence should be called out in the c'tor description.

Reply:
I disagree, I think `new` should cover the most common use case, which I think is what's now `from_headers`. But I don't insist of having it set up that way, this should be just as easy to use.

Reply:
But this way the constructors are consistent across the three kinds of jars:
- Empty jar: `CookieJar::new()`, `SignedCookieJar::new(key)`, `PrivateCookieJar::new(key)`
- With headers: `CookieJar::from_headers(headers)`, `SignedCookieJar::from_headers(headers, key)`, `PrivateCookieJar::from_headers(headers, key)`

Private and signed jars cannot implement `Default` because they need a `Key`.

Reply:
Right, I guess that is a good reason to do it this way 👍